src/server/pkg/worker/master.go

```diff
@@ -0,0 +1,691 @@
+package worker
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"go.pedge.io/lion/proto"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/limit"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pkg/grpcutil"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/dlock"
+	"github.com/pachyderm/pachyderm/src/server/pkg/hashtree"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	pfs_sync "github.com/pachyderm/pachyderm/src/server/pkg/sync"
+)
+
+const (
+	// maximumRetriesPerDatum is the maximum number of times each datum
+	// can failed to be processed before we declare that the job has failed.
+	maximumRetriesPerDatum = 3
+
+	masterLockPath = "_master_worker_lock"
+)
+
+func (a *APIServer) master() {
+	backoff.RetryNotify(func() error {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		masterLock, err := dlock.NewDLock(ctx, a.etcdClient, path.Join(a.etcdPrefix, masterLockPath, a.pipelineInfo.Pipeline.Name))
+		if err != nil {
+			return err
+		}
+		defer masterLock.Unlock()
+		ctx = masterLock.Context()
+
+		protolion.Infof("Launching worker master process")
+
+		// Set pipeline state to running
+		_, err = col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			pipelineName := a.pipelineInfo.Pipeline.Name
+			pipelines := a.pipelines.ReadWrite(stm)
+			pipelineInfo := new(pps.PipelineInfo)
+			if err := pipelines.Get(pipelineName, pipelineInfo); err != nil {
+				return err
+			}
+			pipelineInfo.State = pps.PipelineState_PIPELINE_RUNNING
+			pipelines.Put(pipelineName, pipelineInfo)
+			return nil
+		})
+
+		jobCh := make(chan *pps.JobInfo)
+		defer close(jobCh)
+		go a.jobManager(ctx, jobCh)
+
+		return a.jobSpawner(ctx, jobCh)
```

Comment:
This is another case where we're using channels and concurrency and I'm not at all clear why. Am I correct in thinking that this code works approximately like so:

- create a `branchSetFactory`
- for each set of branches it produces spawn a job
- run that job until completion
- repeat

If so it doesn't seem like there should be any need for multiple goroutines or channels, all you need to do is iterate the results from the `branchSetFactory` and run the `jobManager` for each of them.